### PR TITLE
[WIP] Add support for Bitbucket Cloud

### PR DESCRIPTION
**Pull request by itiut**
Add support for Bitbucket Cloud. Refs #22.
I confirmed that this works well on Chrome, Safari, and Firefox.

For now, this doesn't support for Bitbucket Server and may not work well for Bitbucket Cloud with Mercurial. (I'm sorry but I don't use Bitbucket Server and Mercurial.)

I have two concerns about implementation.

Firstly, there are [app passwords](https://blog.bitbucket.org/2016/06/06/app-passwords-bitbucket-cloud/) on Bitbucket, which are about the same functions as personal access tokens on GitHub, but app passwords can be used only for Basic Authentication. So the username is also required to access private repositories. Get the username embedded in HTML when the user is logged in, and use it for HTTP requests.

Secondly, git submodules can't be retrieved with Bitbucket API [src Resources](https://confluence.atlassian.com/bitbucket/src-resources-296095214.html) but can only by reading `.gitmodules` content. So when reading `.gitmodules` content for the first time, memoize it so that submodules will be inserted into the tree later.

And I have a question about CSS on Safari. Injected styles don't override Bitbucket original styles and the layout is skewed.

If there are something to be fixed, please point out them.

### CODE REVIEW

**Comment by buunguyen** [comment]
This is great, thank you! I'm a bit swamped right now, I'll provide feedback later.

File: src/adapters/adapter.js [hunk 1]
```diff
@@ -287,4 +287,12 @@ class Adapter {
   _getSubmodules(tree, opts, cb) {
     throw new Error('Not implemented')
   }
+
+  /**
+   * Returns item's href value.
+   * @api public
```
Comment (buunguyen) [inline src/adapters/adapter.js]: should be @protected?

File: src/adapters/bitbucket.js [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        container: $pjaxContainer
+      })
+    }
+    else { // falls back
+      super.selectFile(path)
+    }
+  }
+
+  // @override
+  downloadFile(path, fileName) {
```
Comment (buunguyen) [inline src/adapters/bitbucket.js]: This looks like the implementation in base adapter, why redefining?

File: src/adapters/bitbucket.js [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    super(['jquery.pjax.js'])
+
+    $.pjax.defaults.timeout = 0 // no timeout
+    $(document)
+      .on('pjax:send', () => $(document).trigger(EVENT.REQ_START))
+      .on('pjax:end', () => $(document).trigger(EVENT.REQ_END))
+  }
+
+  // @override
+  init($sidebar) {
```
Comment (buunguyen) [inline src/adapters/bitbucket.js]: This method & selectFile() duplicate with github adapter a lot, maybe add a PjaxAdapter to keep the share code?

File: src/template.html [hunk 4]
```diff
@@ -1,6 +1,6 @@
 <div>
   <nav class="octotree_sidebar">
-    <a class="octotree_toggle btn">
+    <a class="octotree_toggle btn aui-button">
```
Comment (buunguyen) [inline src/template.html]: This is too bb-specific, I'd prefer we style in bitbucket.less

File: src/adapters/bitbucket.js [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+  _get(path, opts, cb) {
+    const host = location.protocol + '//' + 'api.bitbucket.org/1.0'
+    const url = `${host}/repositories/${opts.repo.username}/${opts.repo.reponame}${path || ''}`
+    const cfg  = { url, method: 'GET', cache: false }
+
+    if (opts.token) {
+      // Bitbucket App passwords can be used only for Basic Authentication.
+      // Get username of logged-in user.
+      const currentUser = JSON.parse($('body').attr('data-current-user'))
```
Comment (buunguyen) [inline src/adapters/bitbucket.js]: Maybe it's more reliable if we allow user to specify both username & password. We can use a single field a separate by `:`. Just change the verbiage in the case of BitBucket so that it's clear.

In other words, if the token has `:`, assume users enter both and use this username. If users don't explicitly specify username, retrieve it from the page (like currently). If can't retrieve (DOM changes), show error and ask user to enter both username & password, separated by `:`.

File: src/octotree.js [hunk 6]
```diff
@@ -11,13 +11,26 @@ $(document).ready(() => {
   }
 
   function createAdapter() {
+    const normailzeUrl = (url) => url.replace(/(.*?:\/\/[^/]+)(.*)/, '$1')
```
Comment (buunguyen) [inline src/octotree.js]: typo, normalizeUrl

**Comment by buunguyen** [comment]
@itiut did you have a chance to continue working on this?
